runtime/oti/VMHelpers.hpp

```diff
@@ -2070,57 +2089,87 @@ class VM_VMHelpers
+
+		return (complementGCConcurrentState == returnedState);
 	}
 
+	/**
+	 * clear CONCURRENTSCANNING flag bit0:for LocalConcurrentScanning /bit1:for GlobalConcurrentScanning base on checkConcurrentState,
+	 * if all CONCURRENTSCANNING bits(bit0 and bit1) are cleared and the continuation mounting is blocked by concurrent scanning, notify it.
+	 * @param [in] checkConcurrentState can be J9_GC_CONTINUATION_STATE_CONCURRENT_SCAN_LOCAL or J9_GC_CONTINUATION_STATE_CONCURRENT_SCAN_GLOBAL
+	 */
 	static VMINLINE void
-	exitConcurrentGCScan(J9VMContinuation *continuation)
-	{
-		/* clear CONCURRENTSCANNING flag */
-		uintptr_t oldContinuationState = VM_AtomicSupport::bitAnd(&continuation->state, ~(uintptr_t)J9_GC_CONTINUATION_STATE_CONCURRENT_SCAN);
-		J9VMThread *carrierThread = getCarrierThreadFromContinuationState(oldContinuationState);
-		if (NULL != carrierThread) {
-			omrthread_monitor_enter(carrierThread->publicFlagsMutex);
-			/*  notify the waiting carrierThread that we just finished scanning, and it can proceed with mounting. */
-			omrthread_monitor_notify_all(carrierThread->publicFlagsMutex);
-			omrthread_monitor_exit(carrierThread->publicFlagsMutex);
+	exitConcurrentGCScan(J9VMContinuation *continuation, bool isGlobalGC)
+	{
+		/* clear CONCURRENTSCANNING flag bit0:LocalConcurrentScanning /bit1:GlobalConcurrentScanning */
+		uintptr_t oldContinuationState = VM_AtomicSupport::bitAnd(&continuation->state, ~getGCConcurrentState(isGlobalGC));
+		uintptr_t complementGCConcurrentState = oldContinuationState & getGCConcurrentState(!isGlobalGC);
+		if (!complementGCConcurrentState) {
+			J9VMThread *carrierThread = getCarrierThreadFromContinuationState(oldContinuationState);
+			if (NULL != carrierThread) {
+				omrthread_monitor_enter(carrierThread->publicFlagsMutex);
+				/*  notify the waiting carrierThread that we just finished scanning, and it can proceed with mounting. */
+				omrthread_monitor_notify_all(carrierThread->publicFlagsMutex);
+				omrthread_monitor_exit(carrierThread->publicFlagsMutex);
+			}
 		}
 	}
+
 #endif /* JAVA_SPEC_VERSION >= 19 */
 
 	static VMINLINE UDATA
-	walkContinuationStackFramesWrapper(J9VMThread *vmThread, j9object_t continuationObject, J9StackWalkState *walkState, bool syncWithContinuationMounting)
+	walkContinuationStackFramesWrapper(J9VMThread *vmThread, j9object_t continuationObject, J9StackWalkState *walkState, bool isConcurrentGC, bool isGlobalGC)
 	{
 		UDATA rc = J9_STACKWALK_RC_NONE;
 #if JAVA_SPEC_VERSION >= 19
 		J9VMContinuation *continuation = J9VMJDKINTERNALVMCONTINUATION_VMREF(vmThread, continuationObject);
-		if (syncWithContinuationMounting && (NULL != continuation)) {
-			if (!tryWinningConcurrentGCScan(continuation)) {
+		if (isConcurrentGC && (NULL != continuation)) {
+			if (!tryWinningConcurrentGCScan(continuation, isGlobalGC)) {
 				/* If continuation is mounted or already being scanned by another GC thread, we do nothing */
```

Comment:
adjust this comment:

if continuation is mounted or already being scanned by another GC thread of the same GC type, we do nothing